cpp/src/spatial/lonlat_to_cartesian.cu

```diff
@@ -0,0 +1,149 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <type_traits>
+#include <utility>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cuspatial/error.hpp>
+
+namespace {
+
+using pair_of_columns = std::pair<std::unique_ptr<cudf::column>, std::unique_ptr<cudf::column>>;
+
+constexpr double earth_circumference_km = 40000.0;
+constexpr double earth_circumference_km_per_degree = earth_circumference_km / 360.0;
+constexpr double deg_to_rad = M_PI / 180;
+
+CUDA_HOST_DEVICE_CALLABLE
+double midpoint(double a, double b)
+{
+    return (a + b) / 2;
+}
+
+CUDA_HOST_DEVICE_CALLABLE
+double lon_to_x(double lon, double lat)
+{
+    return lon * earth_circumference_km_per_degree * cos(lat * deg_to_rad);
+};
+
+CUDA_HOST_DEVICE_CALLABLE
+double lat_to_y(double lat)
+{
+    return lat * earth_circumference_km_per_degree;
+};
+
+struct lonlat_to_cartesian_functor
+{
+    template <typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, pair_of_columns>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template <typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, pair_of_columns>
+    operator()(double origin_lon,
+               double origin_lat,
+               cudf::column_view const& input_lon,
+               cudf::column_view const& input_lat,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        auto size = input_lon.size();
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto type = cudf::data_type{ tid };
+        auto output_x = cudf::make_fixed_width_column(type, size, cudf::mask_state::UNALLOCATED, stream, mr);
+        auto output_y = cudf::make_fixed_width_column(type, size, cudf::mask_state::UNALLOCATED, stream, mr);
+
+        auto input_iters = thrust::make_tuple(input_lon.begin<T>(),
+                                              input_lat.begin<T>());
+
+        auto output_iters = thrust::make_tuple(output_x->mutable_view().begin<T>(),
+                                               output_y->mutable_view().begin<T>());
+
+        auto input_zip = thrust::make_zip_iterator(input_iters);
+        auto output_zip = thrust::make_zip_iterator(output_iters);
+
+        auto to_cartesian = [=] __device__ (auto lonlat) {
+            auto lon = thrust::get<0>(lonlat);
+            auto lat = thrust::get<1>(lonlat);
+            return thrust::make_pair(
+                lon_to_x(origin_lon - lon, midpoint(lat, origin_lat)),
```

Comment:
The reason I factored this differently from legacy is that `M_PI / 360` is a misnomer. It at first appears to be an incorrectly implemented degree => radian calculation. However, it can also be understood to be taking the midpoint between lat and origin_lat. I find this to be more intelligible than the previous implementation, but I am not sure _why_ it makes sense to take the midpoint of `lat` and `origin_lat`.

Reply:
```
x = ((origin_lon - lon) / 360) * 40000.0 * cos((origin_lat + lat) * M_PI / 360);
```
vs
```
x = ((origin_lon - lon) / 360) * 40000.0 * cos(((origin_lat + lat) / 2) * (M_PI / 180));
```

Reply:
I'm also confused why we are using `origin - position` rather than `position - origin`. To get from A to B, we use `B - A`. Isn't that the question we're trying to answer here?

_"from the perspective of the the origin (A), where is the destination (B)?"_

Instead we're asking

_"from the perspective of the destination (B), where is the origin (A)?_

which gives us the opposite.

This might make sense if it were reducing a more complex equation, such as intentionally inverting x/y because lon/lat are opposite by nature, but as longitude increases, easting increases, and as latitude increases, northing increases. Wouldn't it then make sense that as longitude increases, X also increased? (An argument could be made for 3d graphics and left/right hand rule, but we are returning 2d coordinates).

Reply:
The signs do seem inverted. As for the origin of the formula, the whole thing is unclear to me. Without knowing what map projection is being used, it will be difficult to figure out. I do note that calculating the length of a secant on a circle results in a half angle, which could account for the PI/360 instead of PI/180...

I think we need @zhangjianting to point us at the original code.